### PR TITLE
Handling whitespace in Lexer.cs

### PR DESCRIPTION
**Pull request by A-Russell23**
## Main Changes
My main goal was to allow the lexer to handle whitespaces, rather than having the transpiler do this. This was a matter of keeping track of whitespace and adding it to the front of every token's word.
I also updated the lexer unit tests to check new lines are being dealt with correctly, and that whitespace is being dealt with correctly.

### Considerations
- I decided to discard any trailing whitespace at the end of the line, otherwise it was being added to the start of the next token. I could have found a way to keep it but seen as whitespace isn't really noticeable unless there's a character after it, I didn't think it was worth it.
- I changed tokens that are strings or numbers so that their word contains any whitespace + the characters, and their value contains only the characters. I then changed the parser to look at the value rather than the word when checking for keywords. This needed to be done as the parser was checking the whitespace when looking for keywords, so '  var' would be checked for example, rather than just 'var'. This caused an error when it shouldn't. This would also need to be taken into consideration when using variable identifiers. The value would need to be checked as to not account for the whitespace. I could have taken a different approach to this (ignoring the whitespace in the parser or assigning word and value the other way around), but I felt this was the most appropriate and easy to implement option, however this should be discussed by the team to ensure we choose the right option.

## Bug Fix
- In Lexer.cs the '==' token was being set to 'NotEqual'

### CODE REVIEW

File: C_Flat_Interpreter/Lexer/Lexer.cs [hunk 1]
```diff
@@ -50,124 +50,127 @@ public int Tokenise(string input)
...
+                var newToken = new Token
+                {
+                    Line = j
+                };
 
                 switch (c)
                 {
                     case ' ':
-                        //ignore whitespace
+                        whitespace += " ";
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Lexer/Lexer.cs:61]: Should probably just do `whitespace += c;` instead for consistency :)

Comment (KThompson-Lane) [inline C_Flat_Interpreter/Lexer/Lexer.cs:61]: Could probably just do `whitespace += c;` instead for consistency :)

File: C_Flat_Interpreter/Lexer/Lexer.cs [hunk 2]
```diff
@@ -50,124 +50,127 @@ public int Tokenise(string input)
...
-                            newToken.Word = c.ToString() + _lines[j][i].ToString();
+                            newToken.Type = TokenType.Equals;
+                            newToken.Word = whitespace + c + _lines[j][i];
                         }
                         else
                         {
                             newToken.Type = TokenType.Assignment;
-                            newToken.Word = c.ToString();
+                            newToken.Word = whitespace + c;
                         }
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Lexer/Lexer.cs:125]: Minor thing but maybe remove empty line here for consistent formatting :p

File: C_Flat_Interpreter/Lexer/Lexer.cs [hunk 3]
```diff
@@ -50,124 +50,127 @@ public int Tokenise(string input)
...
-
                         break;
                 }
 
-                if (newToken != null)
-                {
-                    _tokens.Add(newToken);
-                }
+                if (newToken == null) continue;
```
Comment (KThompson-Lane) [inline C_Flat_Interpreter/Lexer/Lexer.cs:169]: Nice job on the continue here 👍

**Review by KThompson-Lane — COMMENTED** [review]
Overall looks really good and I'm happy with the way you've handled the whitespace lexing. Only thing I want to discuss is the choice of using word and trimming whitespaces or keeping the value field around.  

I really do think that value is pretty redundant and having it just be a copy of the word but without the whitespaces doesn't really provide it enough reason to exist. I'm not completely stuck on either method and happy to discuss but at least for me personally, removing value completely and just trimming off the whitespaces when we display the token seems like a more true-to-form lexing implementation.

Not giving explicit approval for now so we can align on a final decision. Good work though 😄

**Comment by A-Russell23** [comment]
> Overall looks really good and I'm happy with the way you've handled the whitespace lexing. Only thing I want to discuss is the choice of using word and trimming whitespaces or keeping the value field around.
> 
> I really do think that value is pretty redundant and having it just be a copy of the word but without the whitespaces doesn't really provide it enough reason to exist. I'm not completely stuck on either method and happy to discuss but at least for me personally, removing value completely and just trimming off the whitespaces when we display the token seems like a more true-to-form lexing implementation.
> 
> A quick example of this implementation would be setting the `newToken.Word` to the complete lexed string with the prepended whitespaces. Then in `Parser` wherever you've now used `token.Value` for strings/numbers you can instead just use the trimmed word e.g. `var trimmedString = token.Word.Trim()` [1](#user-content-fn-note-c2aa28622ac5059c168471e9f2f32c7a)
> 
> Not giving explicit approval for now so we can align on a final decision. Good work though 😄
> 
> ## Footnotes
> 1. The example I gave might not be syntactically correct but it will _hopefully_ match up with the actual method. [↩](#user-content-fnref-note-c2aa28622ac5059c168471e9f2f32c7a)

I think this makes a lot more sense than using value! And yes, value itself is pretty redundant now. I think we should just get rid of it

File: C_Flat_Interpreter/Lexer/Lexer.cs [hunk 4]
```diff
@@ -195,7 +195,7 @@ private string ParseNumber(int line, int index)
         }
         return numberString;
     }
-    private string ParseWord(int line, int index)
+    private string ParseString(int line, int index)
```
Comment (A-Russell23) [inline C_Flat_Interpreter/Lexer/Lexer.cs:198]: Changed this as it seemed a bit misleading

**Review by Spartan-Bailey — APPROVED** [review]
